### PR TITLE
Mrc-3620 Add auth0 login method 

### PR DESCRIPTION
**Pull request by LekanAnni**
PR integrate SSO Authentication from Proof of Concept to production code.  

* I implemented SSO logout in this code, it was not in the proof of concept.
* I added a flag indicating login method, default value is false which indicates users will login with formLogin and OAuth2 where flag is set true.
* Unit Tests have been added, unfortunately integration test can not be added. However, we could benefit from browser tests. I suggest browser tests should test oauth2 login when we finally add selenium or e2e infrastructure in the next milestone.

#Testing
* Please replace `oauth2_**` in config.properties with real config, please let me know if you don't have the real config.
* Alternatively, you can see a deployed instance on dev `https://naomi-dev.dide.ic.ac.uk/login`, where oauth2_login_method is enabled.

## Type of version change
_Delete as appropriate_

Major - Minor - Patches - None

## Checklist

- [x] I have incremented version number, or version needs no increment
- [x] The build passed successfully, or failed because of ADR tests

### CODE REVIEW

**Review by EmmaLRussell — COMMENTED** [review]
Looks good to me, though to be honest all the convolutions you have to go through for pac4j always confuse me!

One issue is that, when I browse to Naomi homepage I'm immediately forwarded to the Auth0 login, so I don't seem to have the option to Continue as guest, which we do want to support. What's the plan for supporting guest users?

> unfortunately integration test can not be added. However, we could benefit from browser tests. I suggest browser tests should test oauth2 login when we finally add selenium or e2e infrastructure in the next milestone.

We could probably test hitting the end points as integration tests (though there would be config issues with integration testing both auth types) as we do with the existing auth method (in `SecureIntegrationTest`) but we would also want to avoid hitting Auth0 with our automated tests I guess, in case they banned us! So we'll probably want to use hoverfly or something similar eventually to mimic responses from Auth0. Do we have an epic for adding e2e tests yet?

File: src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/LoginController.kt [hunk 1]
```diff
@@ -7,16 +7,25 @@ import org.springframework.ui.set
...
+    private val request: HttpServletRequest,
+    private val session: Session,
+    appProperties: AppProperties,
+    oauth2State: OAuth2State
+) : OAuth2AuthenticationRedirection(appProperties, oauth2State)
 {
-
     @GetMapping("/login")
-    fun login(model: Model): String
+    fun login(model: Model): Any
```
Comment (Nicolas-Dolan) [inline src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/LoginController.kt:22]: Can we be more specific than `Any` here?

**Review by Nicolas-Dolan — APPROVED** [review]

File: src/app/src/main/kotlin/org/imperial/mrc/hint/security/oauth2/OAuth2State.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+     * During implementation, code should be used to validate
+     * OAuth2 redirect. Code should be encoded and stored in the state,
+     * when a redirect is requested, code should be decoded and
+     * compared against the stored code for validation. should
+     * save guard against CSRF attacks.")
+     *
+     */
+    fun generateCode(): String
+    {
+        return UUID.randomUUID().toString()
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/security/oauth2/OAuth2State.kt:20]: I have added a test for this. src/app/src/test/kotlin/org/imperial/mrc/hint/unit/security/oauth2/OAuth2StateTests.kt

File: src/app/src/main/kotlin/org/imperial/mrc/hint/security/oauth2/OAuth2State.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+{
+    /**
+     * TODO("
+     * During implementation, code should be used to validate
+     * OAuth2 redirect. Code should be encoded and stored in the state,
+     * when a redirect is requested, code should be decoded and
+     * compared against the stored code for validation. should
+     * save guard against CSRF attacks.")
+     *
+     */
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/security/oauth2/OAuth2State.kt:17]: You are right, I don't thick we have a ticket for this, I have now created a ticket https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-3654/Encode-and-save-generated-code-in-the-state

File: src/app/src/main/kotlin/org/imperial/mrc/hint/security/oauth2/OAuth2AuthenticationRedirection.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        val httpHeader = HttpHeaders()
+        httpHeader.location = URI(url.toUriString())
+        return ResponseEntity(httpHeader, HttpStatus.SEE_OTHER)
+    }
+
+    fun oauth2LogoutRedirect(): String
+    {
+        return UriComponentsBuilder
+            .fromHttpUrl("https://${appProperties.oauth2ClientUrl}")
+            .path("/v2/logout")
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/security/oauth2/OAuth2AuthenticationRedirection.kt:39]: This is the logout endpoint provided by auth0.

File: src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/LoginController.kt (hunk 1 above)
Comment (LekanAnni) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/LoginController.kt:22]: > Auth0. Do we have an epic for adding e2e tests yet?

I am not sure who's working on this bit but I will check with Rob, I think it's going into next quarter milestone. Yes, it's tricky to test even something as simple as hitting auth0 endpoint.

Comment (LekanAnni) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/LoginController.kt]: > Can we be more specific than `Any` here?

You are right but `Any` is a powerful alternative as I had like to return String and ResponseEntity.

**Review by EmmaLRussell — COMMENTED** [review]
> This PR has worked exactly according to the spec. 

Sure.. but perhaps the spec wasn't as well thought out as it could have been! 😆 
OK, let's assume we can roll guest user functionality into another branch - it's not mentioned in the spec for mrc-3617 so I guess that will need re-estimating or splitting in two. Can talk about that at today;s planning meeting.

**Review by EmmaLRussell — APPROVED** [review]

**Pull request merged**